File: elixir/plugins/onepassword.ex
# 1Password, through a Connect server, as a sekreto plugin.
#
# A port of typescript/plugins/onepassword.ts, which is canonical.

defmodule Sekreto.Plugins.Onepassword do
  @moduledoc """
  The `onepassword` provider kind.

  A plugin module: nothing under `src/` names it, and a chain that does not
  name this kind never loads it. See docs/design/plugin-providers.md.
  """

  alias Sekreto.Cell
  alias Sekreto.Error
  alias Sekreto.Json
  alias Sekreto.Plugins.Http
  alias Sekreto.Plugins.Httpjson
  alias Sekreto.ProviderSpec
  alias Sekreto.Providers

  @doc """
  The `onepassword` provider kind, as a voxgig/plugin definition.

  The calling project passes it to the constructor, and only then can a
  chain name the kind:

      Sekreto.new(chain, plugins: [Sekreto.Plugins.Onepassword.onepassword()])
  """
  def onepassword, do: Providers.providerplugin("onepassword", &fromspec/1)

  @doc """
  1Password, through a Connect server.

  The item titled `api.token` (titles keep their dots), in the named vault.
  The value is the field with purpose PASSWORD, or the field labelled
  `value`. A vault that cannot be found is an error - config names it, so
  its absence is a broken store, not a missing secret.
  """
  def onepassword(addr, token, vault) do
    cell = Cell.new(:unresolved)
    auth = [{"authorization", "Bearer " <> token}]

    resolvevault = fn useaddr ->
      if "" == vault, do: raise(Error, message: "sekreto: onepassword: no vault")

      res = Httpjson.fetchjson("GET", useaddr <> "/v1/vaults", auth)
      list = Json.asarr(res.body)

      if 200 != res.status or :none == list do
        raise Error,
          message:
            "sekreto: onepassword error: " <> Httpjson.tostr(res.status) <> ": listing vaults"
      end

      found =
        Enum.find(list, fn entry ->
          vault == Json.text(Json.dig(entry, ["id"])) or
            vault == Json.text(Json.dig(entry, ["name"]))
        end)

      if nil == found do
        raise Error, message: "sekreto: onepassword: no vault named " <> vault
      end

      case Json.text(Json.dig(found, ["id"])) do
        :none -> ""
        id -> id
      end
    end

    %{
      lookup: fn name ->
        Sekreto.checkname(name)

        useaddr = Httpjson.trimslash(addr)
        if "" == useaddr, do: raise(Error, message: "sekreto: onepassword: no addr")
        Providers.checkaddr(useaddr)

        id =
          case Cell.get(cell) do
            :unresolved ->
              resolved = resolvevault.(useaddr)
              Cell.put(cell, resolved)
              resolved

            resolved ->
              resolved
          end

        filter = Http.uriescape(~s|title eq "| <> name <> ~s|"|)
        found =
          Httpjson.fetchjson(
            "GET",
            useaddr <> "/v1/vaults/" <> id <> "/items?filter=" <> filter,
            auth
          )

        items = Json.asarr(found.body)

        if 200 != found.status or :none == items do
          raise Error,
            message:
              "sekreto: onepassword error: " <>
                Httpjson.tostr(found.status) <> ": finding " <> name
        end

        if [] == items do
          nil
        else
          itemid =
            case Json.text(Json.dig(hd(items), ["id"])) do
              :none -> ""
              value -> value
            end

          item =
            Httpjson.fetchjson("GET", useaddr <> "/v1/vaults/" <> id <> "/items/" <> itemid, auth)

          if 200 != item.status do
            raise Error,
              message:
                "sekreto: onepassword error: " <>
                  Httpjson.tostr(item.status) <> ": reading " <> name
          end

          fields =
            case Json.asarr(Json.dig(item.body, ["fields"])) do
              :none -> []
              value -> value
            end

          # Two full passes, in order: purpose first, then label.
          chosen =
            Enum.find(fields, fn field ->
              "PASSWORD" == Json.asstr(Json.dig(field, ["purpose"]))
            end) ||
              Enum.find(fields, fn field ->
                "value" == Json.asstr(Json.dig(field, ["label"]))
              end)

          if nil == chosen, do: nil, else: Httpjson.nonone(Json.text(Json.dig(chosen, ["value"])))
        end
      end,
      describe: fn -> "onepassword:" <> vault end
    }
  end

  defp fromspec(%ProviderSpec{} = spec), do: onepassword(spec.addr, spec.token, spec.vault)
end

File: swift/plugins/Onepassword.swift
// The `onepassword` provider kind: 1Password Connect.
//
// A plugin, not a built-in: it opens a socket.
//
// A port of typescript/plugins/onepassword.ts, which is canonical.

import Foundation

import Sekreto

// A SCOPED import: plugin exports a `Json` of its own, and the only name
// this file wants from it is `Definition`.
import struct VoxgigPlugin.Definition

/// 1Password, through a Connect server.
///
/// The item titled `api.token` (titles keep their dots), in the named
/// vault. The value is the field with purpose PASSWORD, or the field
/// labelled `value`. A vault that cannot be found is an error - config
/// names it, so its absence is a broken store, not a missing secret.
public final class OnepasswordProvider: Provider {

  private let addr: String?
  private let token: String?
  private let vault: String?

  private var vaultid: String?

  public init(addr: String? = nil, token: String? = nil, vault: String? = nil) {
    self.addr = addr
    self.token = token
    self.vault = vault
  }

  private func auth() -> Ordered<String> {
    var out = Ordered<String>()
    out["authorization"] = "Bearer \(token ?? "")"
    return out
  }

  private func resolvevault(_ useaddr: String) throws -> String {
    let want = vault ?? ""
    if want.isEmpty { throw SekretoError("sekreto: onepassword: no vault") }

    let res = try fetchjson("GET", "\(useaddr)/v1/vaults", auth())

    guard 200 == res.status, let list = res.body.aslist else {
      throw SekretoError("sekreto: onepassword error: \(res.status): listing vaults")
    }

    for entry in list {
      let id = entry.dig("id").text
      let name = entry.dig("name").text

      if want == id || want == name {
        return id ?? ""
      }
    }

    throw SekretoError("sekreto: onepassword: no vault named \(want)")
  }

  public func lookup(_ name: String) throws -> String? {
    _ = try checkname(name)

    let useaddr = trimslash(addr ?? "")
    if useaddr.isEmpty { throw SekretoError("sekreto: onepassword: no addr") }
    try checkaddr(useaddr)

    if nil == vaultid {
      vaultid = try resolvevault(useaddr)
    }

    let id = vaultid ?? ""

    let filter = uriescape("title eq \"\(name)\"")
    let found = try fetchjson("GET", "\(useaddr)/v1/vaults/\(id)/items?filter=\(filter)", auth())

    guard 200 == found.status, let items = found.body.aslist else {
      throw SekretoError("sekreto: onepassword error: \(found.status): finding \(name)")
    }

    if items.isEmpty { return nil }

    let itemid = items[0].dig("id").text ?? ""
    let item = try fetchjson("GET", "\(useaddr)/v1/vaults/\(id)/items/\(itemid)", auth())

    if 200 != item.status {
      throw SekretoError("sekreto: onepassword error: \(item.status): reading \(name)")
    }

    let fields = item.body.dig("fields").aslist ?? []

    // Two full passes, in order: a password field wins over a field merely
    // labelled `value`.
    for field in fields where "PASSWORD" == field.dig("purpose").asstr {
      return field.dig("value").text
    }

    for field in fields where "value" == field.dig("label").asstr {
      return field.dig("value").text
    }

    return nil
  }

  public func describe() -> String {
    return "onepassword:\(vault ?? "")"
  }
}

/// The kind, as a voxgig/plugin definition.
public let onepassword: Definition = providerplugin("onepassword") { spec in
  OnepasswordProvider(addr: spec.addr, token: spec.token, vault: spec.vault)
}
